Add test plan for kernel_bundle (#455)
KhronosGroup/SYCL-CTS: test_plans/kernel_bundle.asciidoc

This is a test plan cover kernel_bundle functionality that is not covered by current tests.

== Testing scope

Tests intend to cover functions from 4.11. Kernel bundles that are not covered in https://github.com/KhronosGroup/SYCL-CTS/tree/SYCL-2020/tests/kernel_bundle

=== Device coverage

All of the tests described below are performed only on the default device that
is selected on the CTS command line if not stated otherwise.

== Tests

=== Finctions `is_compatible`

==== Tests for `is_compatible(const std::vector<kernel_id> &kernelIds, const device &dev)`

Check that result of function `is_compatible(const std::vector<kernel_id> &kernelIds, const device &dev)`:

* `true` for zero kernels
* `true` for built-in kernel id for the selected device
* Construct a vector of the built-in kernel ids from every root device other than the selected device.
Check that is_compatible returns false for this vector of kernel ids unless the vector is empty.
* for kernels with no kernel attributes:

** `true` for a kernel that uses `sycl::half` if and only if the selected device has `aspect::fp16`.
** `true` for a kernel that uses `double` if and only if the selected device has `aspect::fp64`.

> ** `true` for a kernel that uses 64-bit atomic operations if and only if the selected device has `aspect::atomic64`.

Author: It was removed by mistake, returned.

> * Define kernels with the following values of the `[[sycl::reqd_work_group_size]]` attribute.

Author: It worked, thank you!

Verify that `is_compatible` returns true if and only if the work group size is less than or equal to the value reported by `info::device::max_work_group_size` for the selected device.

** `[[sycl::reqd_work_group_size(8)]]`
** `[[sycl::reqd_work_group_size(16)]]`
** `[[sycl::reqd_work_group_size(4294967295)]]`

> * Define kernels with the following values of the `[[sycl::reqd_sub_group_size]]` attribute.

Author: Done.

> Verify that `is_compatible` returns true if and only if the sub group size is not one of the values reported by `info::device::sub_group_sizes` for the selected device.

Author: Fixed.

** Check with vector with built-in kernel ids and with the selected device - should return `true`
** To previous vector add kernel id with large `[[sycl::reqd_work_group_size(4294967295)]]` attribute and with the selected device - should return `false` if value reported by `info::device::max_work_group_size` is less than 4294967295.

==== Tests for `template<typename KernelName> bool is_compatible(const device &dev)`

Same tests as previous section except no zero, built-in and multiple kernels cases.

=== `sycl::kernel_bundle` API

==== Deleted default constructor

Check that `std::is_default_constructible_v<sycl::kernel_bundle>` is `false`.

==== Test for `get_backend()`

Call function `kernel_bundle::get_backend()` and check that return type is `sycl::backend`.

==== Tests for `get_devices()`

* Create a 1-element `std::vector<device>` containing the selected device.
* Create kernel_bundle via `get_kernel_bundle<bundle_state::executable>(const context&, const std::vector<device>&)`.
* Call `get_devices()`.
* Check that return type is std::vector<device>.
* Check that return vector contains only selected device.

==== `has_kernel(const kernel_id &kernelId, const device &dev)`

> * Create a kernel bundle in executable state which contains all the application's kernels by calling `get_kernel_bundle<bundle_state::executable>(const context &)`.

Author: But according to [4.11.12.1. Queries](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:bundles.query) `has_kernel(const kernel_id& kernelId, const device& dev)` 
> Returns: true only if the kernel bundle contains the kernel identified by `kernelId` and if that kernel is compatible with the device `dev`.

So in case of not compatible kernel it returns `false` and `is_compatible` also returns `false`, and if test checks condition "`has_kernel(kernelId, dev)` returns `true` if and only if `is_compatible` is `true`" the test are going to pass when they are both `false`. But with ids of kernels that are compatible with selected device they both would return `true` and test will pass. Am I missing something here?

> * Create a kernel bundle in executable state which contains all the application's kernels by calling `get_kernel_bundle<bundle_state::executable>(const context &)`.

Reviewer: Oh, yes you are right.  I had forgotten that `has_kernel(..., dev)` had that semantic.  The test plan you have is good.

One minor comment is that the statement "which contains all the application's kernels" is not correct, as we discussed in the other comment.

> * Create a kernel bundle in executable state which contains all the application's kernels by calling `get_kernel_bundle<bundle_state::executable>(const context &)`.

Author: Thank you! Fixed.

* Iterate over all of the kernel ids returned by `get_kernel_ids()`.
* Verify that `has_kernel(kernelId, dev)` returns `true` if and only if `is_compatible({kernelId}, const device&)` for selected device is `true`.

==== `std::vector<kernel_id> get_kernel_ids()`

Define 4 kernels.

> Get kernel_bundle that contains all of the kernels defined in the application via `get_kernel_bundle<bundle_state::executable>(const context &)`

Author: Thank you! Fixed discription.

Check that result of `get_kernel_ids`:

* has type `std::vector<kernel_id>`
* has size at least 4
* has only different kernel ids

==== `get_kernel()`

* Define kernel KernelName.
* Get a kernel bundle in executable state that contains this kernel via `get_kernel_bundle<KernelName, bundle_state::executable>(const context &)`.
* Use `get_kernel<KernelName>()` to get kernel and check that return type is `sycl::kernel`.

=== Tests for working with specialization constants

Partially tested in https://github.com/KhronosGroup/SYCL-CTS/blob/SYCL-2020/tests/specialization_constants/specialization_constants_via_kernel_bundle.h

> There are should be defined spec constant `SpecName`.

Reviewer: ```suggestion
There are two spec constant defined: `SpecName` and `OtherSpecName`.
```

The test below says that there is also a spec constant named `OtherSpecName`.

Author: Returned `OtherSpecName`.

* Check that `contains_specialization_constants()` return `false`.
* Check that `native_specialization_constant()` return `false`.
* Check that `has_specialization_constant<SpecName>()` return `false`.

==== Kernel bundle with `kernel_handler::get_specialization_constant()` call

* Define a kernel named `KernelName` that calls `kernel_handler::get_specialization_constant<SpecName>()`.
* Attempt to get a kernel bundle in input state that contains this kernel by calling `get_kernel_bundle<KernelName, bundle_state::input>(const context &, const std::vector<device>& )` with 1-element `std::vector<device>` containing the selected device.
* Test if the kernel bundle contains that kernel by calling `kernel_bundle::has_kernel<KernelName()`. If this returns `false`, the test is skipped.
* Set each spec constant to a different value via `kernel_bundle::set_specialization_constant()`.
* Check that `contains_specialization_constants()` return `true`.
* Check that called `native_specialization_constant()` without exception.
* Check that `has_specialization_constant<SpecName>()` return `true`.

> * Check that `has_specialization_constant<OtherSpecName>()` return `false`.

Author: Added this requirement to discription of spec constants above

:sectnums:
:xrefstyle: short

= Test plan for kernel_bundle (cover gaps)

This is a test plan cover kernel_bundle functionality that is not covered by current tests.

== Testing scope

Tests intend to cover functions from 4.11. Kernel bundles that are not covered in https://github.com/KhronosGroup/SYCL-CTS/tree/SYCL-2020/tests/kernel_bundle

=== Device coverage

All of the tests described below are performed only on the default device that
is selected on the CTS command line if not stated otherwise.

== Tests

=== Finctions `is_compatible`

==== Tests for `is_compatible(const std::vector<kernel_id> &kernelIds, const device &dev)`

Check that result of function `is_compatible(const std::vector<kernel_id> &kernelIds, const device &dev)`:

* `false` for zero kernels
* `true` for built-in kernel id for the same device
* `false` for built-in kernel id for the other device
* for kernel with no kernel attributes:

** `true` for kernel with no optional features and any device
** `true` for fp16/fp64/atomic64 kernel and device with appropriate aspect
** `false` for fp16/fp64/atomic64 kernel and device without appropriate aspect

* for kernel with `reqd_work_group_size`

> ** `false` for kernel with no optional features and any device with `max_work_group_size` less than `reqd_work_group_size`

Author: Added, for sizes 8, 16 and 4294967295. And for `info::device::max_work_group_size`.
Values were selected similarly to optional kernel features tests.

* for kernel with `reqd_sub_group_size` (that is less than `max_work_group_size`)

** `false` for kernel with no optional features and any device with `sub_group_sizes` not containing `reqd_sub_group_size`

> * for kernel with no optional features used and with sycl::requires() kernel attribute

Author: Fixed.

** `true` for `has(fp16)`/`has(fp64)`/`has(atomic64)` and device with appropriate aspect

> ** false for `has(fp16)`/`has(fp64)`/`has(atomic64)` and device without appropriate aspect

Author: Done.

* for multiple kernels result is equal to AND operation applied to the multiple calls - each one using single kernel id

> ** Check with vector with kernel ids for same and other devices

Author: Removed other devices and added clarification.

==== Tests for `template<typename KernelName> bool is_compatible(const device &dev)`

Same tests as previous section except no zero, built-in and multiple kernels cases.

=== `sycl::kernel_bundle` API

==== Deleted default constructor

Check that `std::is_default_constructible_v<sycl::kernel_bundle>` is `false`.

==== Test for `get_backend()`

> Call function `kernel::get_backend()` and check that return type is `sycl::backend`.

Author: Fixed.

==== Tests for `get_devices()`

> * Use some of available devices to create fill std::vector<device>.

Author: Done.

> * Create kernel_bundle via `get_kernel_bundle(const context&, const std::vector<device>&)`.

Author: Added.

* Check that return type is std::vector<device>.
* Check that return vector contains all devices from input vector and only them.

==== `has_kernel(const kernel_id &kernelId, const device &dev)`

Check that result of function is:

* `true` if kernelId is within get_kernel_ids() and `is_compatible({kernelId}, const device&)` is `true`
* `false` if kernelId is within get_kernel_ids() and `is_compatible({kernelId}, const device&)` is `false`
* `false` if kernelId is not within get_kernel_ids() and `is_compatible({kernelId}, const device&)` is `true`

> * `false` if kernelId is not within get_kernel_ids() and `is_compatible({kernelId}, const device&)` is `false`

Author: Changed to creating kernel bundle with get_kernel_bundle<bundle_state::executable>(const context &) to possibly get kernels that is incompatible with selected device too.

`has_kernel(const kernel_id &)` without the device parameter is tested in cases for other functionality, for example in tests for `get_kernel_bundle` and I think adding it here would just dupilcate checks there.

==== `std::vector<kernel_id> get_kernel_ids()`

Define 4 kernels.

> Get kernel_bundle with built-in kernels.

Author: Yes, changed to kernel bundle that contains all of the kernels defined in the application and check that the size is at least 4.

Check that result of get_kernel_ids:

* has type `std::vector<kernel_id>`
* has size 4
* has only different kernel ids

==== `get_kernel()`

Define kernel KernelName.

> Get `kernel_bundle` with built-in kernels.

Reviewer: I think you probably want to create the kernel bundle like this:

```suggestion
Get a kernel bundle in executable state that contains this kernel via `get_kernel_bundle<KernelName, bundle_state::executable>(const context &, const device &)`.
```

Author: Changed to "Get a kernel bundle in executable state that contains this kernel via `get_kernel_bundle<KernelName, bundle_state::executable>(const context &)`."

* Use `get_kernel<KernelName>()` to get kernel and that return type is `sycl::kernel`

> * Try to use `get_kernel<OtherKernelName>()` and check that `errc::invalid` is thrown

Reviewer: This part of the test is not valid.  The kernel bundle could contain other kernels also, so it might contain `OtherKernelName`.

Author: You are right, removed.

=== Tests for working with specialization constants

Partially tested in https://github.com/KhronosGroup/SYCL-CTS/blob/SYCL-2020/tests/specialization_constants/specialization_constants_via_kernel_bundle.h

There are should be two spec constants defined - `SpecName` and `OtherSpecName`.

==== Empty kernel bundle

> * Get an empty `kernel_bundle`.

Reviewer: I think there is only one way to get a kernel bundle that is guaranteed to be empty, and it is not obvious, so you should describe it here.

> Get an empty kernel bundle by calling `get_kernel_bundle<bundle_state::executable>(const context &, const std::vector<device> &, Selector)` where `Selector` is a function that always returns `false`.

Also, which bundle state do you expect this test to use?  My wording above assumes you want `executable` state, but I'm not sure if that's what you intend.

Author: Done.

I was thinking to test it only with one state and since legitimacy if this function with `bundle_state::input` and `bundle_state::object` depends on device aspects it seems it would be OK to check only `executable` state.

* Check that `contains_specialization_constants()` return `false`.
* Check that `native_specialization_constant()` return `false`.
* Check that `has_specialization_constant<SpecName>()` return `false`.

> ==== Kernel bundle without `kernel_handler::get_specialization_constant()` call

Reviewer: I think this entire test is not valid because the kernel bundle could contain other kernels in the application, and those other kernels might have specialization constants.

Author: Removed.

* Get an input `kernel_bundle` with kernel without `kernel_handler::get_specialization_constant()` call.
* Check that `contains_specialization_constants()` return `false`.
* Check that `native_specialization_constant()` return `false`.
* Check that `has_specialization_constant<SpecName>()` return `false`.
* Check that `get_specialization_constant<SpecName>()` return default value.
* Call `compile()` to build the `kernel_bundle` into `object` state.
* Check the same.
* Call `link()` to build the `kernel_bundle` into `executable` state.
* Check the same.

==== Kernel bundle with `kernel_handler::get_specialization_constant()` call

> * Get an input `kernel_bundle` with kernel with `kernel_handler::get_specialization_constant<SpecName>()` call.

Reviewer: There is no guarantee that the kernel is available in `input` state, so the test needs to check for this.  Something like:

> * Define a kernel named `KernelName` that calls `kernel_handler::get_specialization_constant<SpecName>()`.
> * Attempt to get a kernel bundle in `input` state that contains this kernel by calling `get_kernel_bundle<KernelName, bundle_state::input>(const context &, const device &)`.
> * Test if the kernel bundle contains that kernel by calling `kernel_bundle::has_kernel<KernelName()`.  If this returns `false`, the test is skipped.

Author: Thank you! Changed to obtaining `kernel_bundle` via `get_kernel_bundle<KernelName, bundle_state::input>(const context &, const std::vector<device>& )` with 1-element `std::vector<device>` containing the selected device.
